concurrent/futures/_base.py

```diff
@@ -194,16 +217,19 @@ def as_completed(fs, timeout=None):
         end_time = timeout + time.time()
 
     fs = set(fs)
+    total_futures = len(fs)
     with _AcquireFutures(fs):
         finished = set(
                 f for f in fs
                 if f._state in [CANCELLED_AND_NOTIFIED, FINISHED])
         pending = fs - finished
         waiter = _create_and_install_waiters(fs, _AS_COMPLETED)
-
+    finished = list(finished)
     try:
-        for future in finished:
-            yield future
+        for f in _yield_finished_futures(finished, waiter,
+                                         ref_collect=(fs,)):
+            f = [f]
```

Comment:
Why this? Why not just `yield f`?

Reply:
Ahhh...because otherwise the future would still be referenced by the generator. Gotcha.

Reply:
Indeed. Maybe Python 4 will consider fixing iteration target vars leaking to the scope :-).